plugins/field-dependent-dropdown/src/field_dependent_dropdown.ts

```diff
@@ -0,0 +1,354 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview A tool for making a dropdown field's options depend on the
+ * value of another field.
+ */
+
+import * as Blockly from 'blockly/core';
+import {Block} from 'blockly/core/block';
+import {FieldDropdown, MenuOption, MenuGeneratorFunction} from 'blockly/core/field_dropdown';
+import {FieldValidator} from 'blockly/core/field';
+
+/**
+ * A deep equality comparison between the two provided arrays, recursively
+ * comparing any child elements that are also arrays.
+ *
+ * @param a The first array to compare.
+ * @param b The second array to compare.
+ * @returns Whether the arrays are deeply equivalent.
+ */
+function arraysAreEquivalent<T>(a: T[], b: T[]): boolean {
+  return a.length === b.length && a.every((aElement, index) => {
+    const bElement = b[index];
+    if (Array.isArray(aElement) && Array.isArray(bElement)) {
+      return arraysAreEquivalent(aElement, bElement);
+    }
+    return aElement === bElement;
+  });
+}
+
+/** The structure of a serialized DependentDropdownOptionsChange. */
+export interface DependentDropdownOptionsChangeJson
+    extends Blockly.Events.BlockBaseJson {
+  name: string;
+  newValue: string;
+  oldValue: string;
+  oldOptions: MenuOption[];
+  newOptions: MenuOption[];
+}
+
+/** The type of the mapping from parent value to child options. */
+export interface ChildOptionMapping{
+  [key: string]: MenuOption[];
+}
+
+declare module 'blockly/core/field_dropdown' {
```

Comment:
This doesn't feel right to me... I don't think we should be reaching into Blockly core's namespace and adjusting it. We would normally have a subclass of the Field to do this. Or you could define your own interface.

Reply:
Okay, I've redesigned the plugin as a subclass of FieldDropdown, and I think it turned out to be a much better API! It no longer requires using the JavaScript interface!